readthedocs/oauth/services/github.py

```diff
@@ -277,39 +280,46 @@ def update_webhook(self, project, integration):
         :rtype: (Bool, Response)
         """
         session = self.get_session()
-        integration.recreate_secret()
+        if not integration.secret:
+            integration.recreate_secret()
         data = self.get_webhook_data(project, integration)
         resp = None
+
+        # Handle the case where we don't have a proper provider_data set
+        # This happens with a user-managed webhook previously
+        if not integration.provider_data:
+            return self.setup_webhook(project, integration)
```

Comment:
I think it's probably better here to actually query the webhook via GitHub: https://developer.github.com/v3/repos/hooks/#list-hooks -- we can check for the `config['url']` and then map that to our existing one. This lets us not re-create webhooks users have already created, but instead take over control of it.